File: v2/src/main/java/com/api/v2/customers/utils/CustomerFinderUtil.java
package com.api.v2.customers.utils;

import com.api.v2.customers.exceptions.NonExistentCustomerException;
import com.api.v2.customers.domain.Customer;
import com.api.v2.customers.domain.CustomerRepository;
import com.api.v2.people.utils.PersonFinderUtil;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

@Component
public class CustomerFinderUtil {

    private final PersonFinderUtil personFinderUtil;
    private final CustomerRepository customerRepository;

    public CustomerFinderUtil(
            PersonFinderUtil personFinderUtil,
            CustomerRepository customerRepository
    ) {
        this.personFinderUtil = personFinderUtil;
        this.customerRepository = customerRepository;
    }

    public Mono<Customer> findBySsn(String ssn) {
        return personFinderUtil
                .findBySsn(ssn)
                .singleOptional()
                .flatMap(optional -> {
                   if (optional.isEmpty()) {
                       return Mono.error(new NonExistentCustomerException(ssn));
                   }
                   return customerRepository
                           .findAll()
                           .filter(c -> c.getPerson().equals(optional.get()))
                           .single();
                });
    }
}

File: v2/src/main/kotlin/com/api/v2/customers/utils/CustomerFinderUtil.kt
package com.api.v2.customers.utils

import com.api.v2.customers.domain.Customer
import com.api.v2.customers.domain.CustomerRepository
import com.api.v2.people.utils.PersonFinderUtil
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.single
import kotlinx.coroutines.withContext
import org.springframework.stereotype.Component

@Component
class CustomerFinderUtil(
    private val personFinderUtil: PersonFinderUtil,
    private val customerRepository: CustomerRepository
) {

    suspend fun findBySsn(ssn: String): Customer {
        return withContext(Dispatchers.IO) {
            val foundPerson = personFinderUtil.findBySsn(ssn)
            println(foundPerson)
            customerRepository
                .findAll()
                .filter { c -> c.person.id == foundPerson.id }
                .single()
        }
    }

}